### PR TITLE
New HTTP and Slack example

### PR DESCRIPTION
**Pull request by pashafateev**

### CODE REVIEW

File: http_slack/program.star [hunk 1]
```diff
@@ -0,0 +1,70 @@
+""" This program illustrates Autokitteh's usage with Slack and HTTP
```
Comment (daabr) [inline http_slack/program.star:1]: The first line in a Python (or Python-like) string should be a short description of the file/module.

Other than the suggestion below, I'd rephrase it to focus on its functionality rather than the integrations it showcases.

```suggestion
"""This program illustrates Autokitteh's usage with Slack and HTTP.
```

File: http_slack/program.star [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+This program queries a server for a list of nodes and posts a message to
+a Slack channel when new nodes are added or removed.
```
Comment (daabr) [inline http_slack/program.star:4]:
```suggestion
a Slack channel when nodes are added or removed.
```

File: http_slack/program.star [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+API details:
+- Web API reference: https://api.slack.com/methods
+- Events API reference: https://api.slack.com/events?filter=Events
+
+This program also demonstrates using a custom builtin function (sleep)
+to sleep for a specified number of seconds.
+
+This program isn't meant to cover all available functions and events.
+It merely showcases various illustrative, annotated, reusable examples.
```
Comment (daabr) [inline http_slack/program.star:14]: These 3 paragraphs are relevant for the sample project that showcases AK's HTTP capabilities. In contrast, this project may assume that the reader is interested in its form and function rather the APIs it uses.

File: http_slack/program.star [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+Starlark is a dialect of Python (see https://bazel.build/rules/language).
+"""
+
+load("@http", "http_no_auth")
+load("@slack", "my_slack")
+load("env", "SERVER")
+load("env", "ENDPOINT")
+load("env", "SLACK_CHANNEL")
```
Comment (daabr) [inline http_slack/program.star:23]: `load` statements can be combined when the source is the same:
```suggestion
load("env", "ENDPOINT", "SERVER", "SLACK_CHANNEL")
```

**Review by daabr — COMMENTED** [review]
Some comments about everything except the code.

File: http_slack/program.star [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            previous_state.remove(node_id)
+        else:
+            new_nodes.append(node_id)
+
+    for node_id in previous_state:
+        removed_nodes.append(node_id)
+
+    return all_nodes, new_nodes, removed_nodes
+
+def on_http_get():
```
Comment (daabr) [inline http_slack/program.star:61]: This is not a universal convention, but I prefer having the "most important" stuff at the top (where one begins reading), and the "least important" stuff at the bottom.

So what is "least" and "most" important here? Top-down and bottom-up are equivalent approaches, neither is universally better than the other. However, arguably, the most impactful part of the code is the function that is triggered by AK, i.e. `on_http_get`, wether or not it was the first piece of code you wrote.

File: http_slack/program.star [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    all_nodes = []
+    while True:
```
Comment (daabr) [inline http_slack/program.star:63]: FYI, no action needed - technically the Starlark spec doesn't **require** `while` to be supported, to prevent endless recursions, but AK does support it.

See https://github.com/google/starlark-go/blob/master/doc/spec.md#while-loops

File: http_slack/program.star [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        all_nodes, new_nodes, removed_nodes = check_node_state(all_nodes)
+        if new_nodes:
+            my_slack.chat_post_message(SLACK_CHANNEL, "New nodes: " + str(new_nodes))
+        if removed_nodes:
+            my_slack.chat_post_message(SLACK_CHANNEL, "Removed nodes: " + str(removed_nodes))
+
+        sleep(10)
```
Comment (daabr) [inline http_slack/program.star:70]: Optional: use a var (with the default set to 10) to allow this duration to be configurable. Tip: the var name should clarify that the unit of measure is seconds (e.g. `INTERVAL_SECS`, `SLEEP_SECS`, etc.)

File: http_slack/program.star [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+load("@slack", "my_slack")
+load("env", "SERVER")
+load("env", "ENDPOINT")
+load("env", "SLACK_CHANNEL")
+
+def get_node_list():
+    node_ids = []
+    next_token = None
+    while True:
+        full_endpoint = SERVER + ENDPOINT
```
Comment (daabr) [inline http_slack/program.star:29]: If this is the only place you use both vars, why not combine them into a single URL?

Unrelated to that, `url` is a more succinct and familiar name than `full_endpoint`.

**Review by daabr — COMMENTED** [review]
A few more tiny comments about the code. All-in-all, a great first project, well done!

File: http_slack/autokitteh.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+project:
+  name: http_sample
+  vars:
+    - name: SERVER
+      value: http://127.0.0.1:63365
+    - name: ENDPOINT
+      value: /api/v1/orchestrator/nodes
+    - name: SLACK_CHANNEL
+      value: "#pasha-slack-test"
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply http_slack/autokitteh.yaml:15]: So the value here can be any arbitrary string? Would it make sense to add a signal to the user like "#your-slack-channel" or something like that?

File: http_slack/program.star (hunk 8 above)
Comment (pashafateev) [reply http_slack/program.star:29]: I was thinking the user would want more than one endpoint. So not that this is useful in the code here but rather as an example.

File: http_slack/autokitteh.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  connections:
+    - name: http_no_auth
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply http_slack/autokitteh.yaml]: Done

File: http_slack/autokitteh.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+      integration: http
+    - name: my_slack
+      integration: slack
+  triggers:
+    - name: http_noauth_get
+      connection: http_no_auth
+      # More HTTP methods are supported as event types:
+      # head, put, delete, options, and patch.
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply http_slack/autokitteh.yaml]: Done

File: http_slack/autokitteh.yaml [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+      event_type: get
+      data:
+        path: get
+      # Triggered by GET request to http://localhost:9980/http/http_sample/get
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply http_slack/autokitteh.yaml]: Done

File: http_slack/autokitteh.yaml [hunk 13]
```diff
@@ -0,0 +1,30 @@
+# This YAML file is a declarative manifest that describes the setup
+# of an AutoKitteh sample project which demonstrates bidirectional
+# integration with HTTP and Slack.
+
+version: v1
+
+project:
+  name: http_sample
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply http_slack/autokitteh.yaml]: Done

File: http_slack/program.star (hunk 1 above)
Comment (pashafateev) [reply http_slack/program.star]: Done

File: http_slack/program.star (hunk 2 above)
Comment (pashafateev) [reply http_slack/program.star]: Done

File: http_slack/program.star (hunk 3 above)
Comment (pashafateev) [reply http_slack/program.star]: Done

File: http_slack/program.star (hunk 4 above)
Comment (pashafateev) [reply http_slack/program.star]: Done

File: http_slack/program.star (hunk 5 above)
Comment (pashafateev) [reply http_slack/program.star]: Done

File: http_slack/program.star (hunk 7 above)
Comment (pashafateev) [reply http_slack/program.star]: Done

File: http_slack/program.star [hunk 14]
```diff
@@ -0,0 +1,59 @@
+"""This program queries a server and posts a message to a Slack channel.
+
+An HTTP GET request triggers this program to query a server for a list 
+of nodes. It then posts a message to a Slack channel when nodes are 
+added or removed.
+
+Starlark is a dialect of Python (see https://bazel.build/rules/language).
+"""
+
+load("@http", "http_no_auth")
```
Comment (itayd) [inline http_slack/program.star:10]: shouldn't this be `my_http`?

File: http_slack/autokitteh.yaml [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+# integration with HTTP and Slack.
+
+version: v1
+
+project:
+  name: http2slack
+  vars:
+    - name: URL
+      value: http://127.0.0.1:63365/api/v1/orchestrator/nodes
+    - name: SLEEP_SECONDS
```
Comment (itayd) [inline http_slack/autokitteh.yaml:12]: it is SLEEP_SECONDS, but the meaning of this is `POLL_INTERVAL`

File: http_slack/program.star [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+load("env", "SLEEP_SECONDS","SLACK_CHANNEL", "URL")
+
+def on_http_get():
+    all_nodes = []
+    while True:
+        all_nodes, new_nodes, removed_nodes = check_node_state(all_nodes)
+        if new_nodes:
+            my_slack.chat_post_message(SLACK_CHANNEL, "New nodes: " + str(new_nodes))
+        if removed_nodes:
+            my_slack.chat_post_message(SLACK_CHANNEL, "Removed nodes: " + str(removed_nodes))
```
Comment (itayd) [inline http_slack/program.star:21]: "nodes left"?

File: http_slack/program.star [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+Starlark is a dialect of Python (see https://bazel.build/rules/language).
+"""
+
+load("@http", "http_no_auth")
+load("@slack", "my_slack")
+load("env", "SLEEP_SECONDS","SLACK_CHANNEL", "URL")
+
+def on_http_get():
+    all_nodes = []
+    while True:
```
Comment (itayd) [inline http_slack/program.star:16]: this will never terminate, i wonder if that's what we want? not sure we have a better solution other than stopping the session via api.

File: http_slack/program.star [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        all_nodes, new_nodes, removed_nodes = check_node_state(all_nodes)
+        if new_nodes:
+            my_slack.chat_post_message(SLACK_CHANNEL, "New nodes: " + str(new_nodes))
+        if removed_nodes:
+            my_slack.chat_post_message(SLACK_CHANNEL, "Removed nodes: " + str(removed_nodes))
+
+        sleep(SLEEP_SECONDS)
+
+def check_node_state(previous_state):
```
Comment (itayd) [inline http_slack/program.star:25]: check_nodes?

File: http_slack/program.star [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    node_ids = []
+    next_token = None
+    while True:
+        url = URL
+        if next_token:
+            url += "?next_token=" + next_token
+        response = http_no_auth.get(url)
+
+        if response.status_code >= 400:
+            print("Error: ", response.status)
```
Comment (itayd) [inline http_slack/program.star:51]: should we terminate the workflow here?

File: http_slack/autokitteh.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (daabr) [reply http_slack/autokitteh.yaml]: sure 👍

File: http_slack/program.star (hunk 18 above)
Comment (daabr) [reply http_slack/program.star:25]: @pashafateev FYI - the notation `s/foo/bar/` comes from CLI tools like `sed` (see https://archive.flossmanuals.net/command-line/sed.html)

Translation from [Leet](https://en.wikipedia.org/wiki/Leet) to English: "Replace `foo` with `bar`".

File: http_slack/program.star (hunk 19 above)
Comment (daabr) [reply http_slack/program.star:51]: My 2 cents: on one hand this sounds like an elegant way to terminate once the server is taken down (e.g. in a demo), but on the other hand it makes this a naive monitoring sample.

Compromise suggestion: maybe send a Slack message instead of `print`, and then end the workflow?

Comment (pashafateev) [reply http_slack/program.star:51]: > My 2 cents: on one hand this sounds like an elegant way to terminate once the server is taken down (e.g. in a demo), but on the other hand it makes this a naive monitoring sample.
> 
> Compromise suggestion: maybe send a Slack message instead of `print`, and then end the workflow?

I like the compromise. Is there a particular way we terminate in the context of Starlark and AK? is `fail()` a good option?

Comment (daabr) [reply http_slack/program.star:51]: no, just patch the code so that the event handler function ends

File: http_slack/program.star (hunk 18 above)
Comment (pashafateev) [reply http_slack/program.star:25]: Done

File: http_slack/program.star [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+            my_slack.chat_post_message(SLACK_CHANNEL, "New nodes: " + str(new_nodes))
+        if removed_nodes:
+            my_slack.chat_post_message(SLACK_CHANNEL, "Removed nodes: " + str(removed_nodes))
+
+        sleep(SLEEP_SECONDS)
+
+def check_node_state(previous_state):
+    new_nodes = []
+    removed_nodes = []
+    all_nodes = get_node_list()
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply http_slack/program.star:28]: Done

File: http_slack/program.star (hunk 19 above)
Comment (pashafateev) [reply http_slack/program.star:51]: Done

File: http_slack/autokitteh.yaml (hunk 15 above)
Comment (pashafateev) [reply http_slack/autokitteh.yaml:12]: Done

File: http_slack/program.star (hunk 14 above)
Comment (pashafateev) [reply http_slack/program.star:10]: Done

File: http_slack/program.star [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+An HTTP GET request triggers this program to query a server for a list 
+of nodes. It then posts a message to a Slack channel when nodes are 
+added or removed.
+
+Starlark is a dialect of Python (see https://bazel.build/rules/language).
+"""
+
+load("@http", "http_no_auth")
+load("@slack", "my_slack")
+load("env", "SLEEP_SECONDS","SLACK_CHANNEL", "URL")
```
[the remark this answers is not in the document]
Comment (pashafateev) [reply http_slack/program.star:12]: Done
